### DISCUSSION TITLE
Creating an AudioSplittingDataset with one Raven file- problem

### DISCUSSION THREAD

**Discussion by canihaveabravo** [kitzeslab/opensoundscape#821] [Q&A]
Hello,

I would really appreciate your help. I am attempting to pre-process a 15-minute audio (.wav) file- split it into 60 s clips, build FFT, augmentation etc and pair the clips with boxed annotations of a antarctic blue whale call derived from a single raven selection table.

I've been using the AudioSplittingDataset function to split the 15-minute files into consecutive 60 s clips and the SpectrogramPreprocessor function to build the FFT and do the augmentation. However I am unable so far to pair the annotations within the dataset.  The tutorial example for AudioSplittingDataset doesn't show how to do this. Please could you provide some instruction for this task?

Thank you very much in advance,

Lorenzo

**Comment by louisfh** [comment]
Thanks for posting! Can you share some of your code and let us know what version of OpenSoundscape you're using so we can see where you're hitting a problem?

I might have misunderstood your question, but there isn't really a one-to-one matching between raven annotations and the output of a preprocessor. We handle raven annotations by turning them into labels for discrete audio segments. So for example a raven file of boxed annotations that looks like this (ignoring extra columns, like frequency, channel etc):
| audio_file | raven_file| start_time        | end_time         | annotation
|:-------:|:-------:|:-------------: |:-------------:|:-------:|
| file1.wav | file1.selections.txt |  5      | 45| beep | 
| file1.wav | file1.selections.txt |   30   | 80 |  boop |
| file1.wav |file1.selections.txt |   190 | 220 |  long_boop |

Would be read in, and then converted to a labels_dataframe
```
from opensoundscape.annotations import BoxedAnnotations
annotations = BoxedAnnotations.from_raven_files(["/path/to/file1.selections.txt"],audio_files=["/path/to/file1.wav"])
labels_dataframe = annotations.one_hot_clip_labels(
    full_duration=240, # The duration of the entire audio file
    clip_duration=60,  # the duration of clip you're interested in
    clip_overlap=0,      # overlap between consecutive clips
    min_label_overlap=0.25, # minimum overlap with a boxed call for an audio segment to be considered to contain that call
)
```


The output, `labels_dataframe` would look like this:
| file | start_time        | end_time         | boop | beep | long_boop |
|:-------:|:-------:|:-------------: |:-------------:|:-------:|:-------:|
| file1.wav | 0 | 60 | 1 | 1 | 0 | 
| file1.wav | 60 |   120   | 1 |  0 | 0 |
| file1.wav |120  | 180 | 0 |  0 | 0 |
| file1.wav |180  | 240 | 0 |  0 | 1 |

It's this labels_dataframe we use to initialize a preprocessor, or train a model. e.g.:
```
from opensoundscape import CNN
model = CNN('resnet18',classes=["boop", "beep", "long_boop"],sample_duration=60.0)
model.train(train_df = labels_dataframe)

**Comment by canihaveabravo** [comment]
Hi Louis,

Thanks very much for your reply and suggestions. I have now managed to process a sample audio file and set up a label_df, preprocess the audio and run a basic test CNN. Some excerpts of code below…

For my next step I would like to import multiple 15-minute audio files and preprocess them to 60 s clips and create a labels dataframe containing multi-label annotations (ABW, noise, OW). The problem I currently have is that these annotations are saved in just one raven selection table (attached). Am I able to create a label dataframe from one raven file that corresponds to annotations covering multiple 15-min audio files (85 file sin this subset?

Thanks for your help, much appreciated.

Lorenzo

# Audio file
audio_file = './ABW/2019-09-04T09-30-06.wav'
# Raven annotation file
annotation_file = './ABW/2019-09-04T09-30-06.ABW.selections.txt'

#create an object from Raven file
annotations = BoxedAnnotations.from_raven_files([annotation_file], keep_extra_columns= None, audio_files=[audio_file])


#inspect the object's .df attribute, which contains the table of annotations
annotations.df.head()
    full_duration=900, # The duration of the entire audio file
    clip_duration=60,
    clip_overlap=0,
    class_subset=None,
    min_label_overlap=0.25,
    final_clip=None
)
labels_df.head()


#initialise preprocessor and set clip length
pre = SpectrogramPreprocessor(sample_duration=60)

# set sample rate on audio load
pre.pipeline.load_audio.set(sample_rate=1000)

#adjust bandpass parameters
pre.pipeline.bandpass.set(min_f=0,max_f=100)

# adjust spectrogram parameters
pre.pipeline.to_spec.params.window_samples = 1024
pre.pipeline.to_spec.params.overlap_fraction = 0.5
pre.pipeline.to_spec.params.decibel_limits = (-140, -10)
pre.pipeline.to_spec.params.dB_scale = True
pre.pipeline.to_spec.params.scaling = "spectrum"

# bypass augmentations
pre.pipeline.random_affine.bypass=False
pre.pipeline.add_noise.bypass =True

# adjust augmentation parameters
pre.pipeline.time_mask.set(max_width = 0.08, max_masks = 5)
pre.pipeline.frequency_mask.set(max_width = 0.05, max_masks = 5)

# Create preprocessed dataset of audio clips from the original audio file
dataset = AudioFileDataset(labels_df, pre)

# Set output tensor shape
dataset.preprocessor.out_shape = [224,224,3]

# Get the first 9 samples and plot them
tensors = [dataset[i].data for i in range(9)]
sample_labels = [list(dataset[i].labels[dataset[i].labels>0].index) for i in range(9)]

_ = show_tensor_grid(tensors,3, labels=sample_labels)

***@***.***

# Create model object
classes = train_df.columns #in this case, there's just one class: ["ABW"]
model = CNN('resnet18',classes=classes,sample_duration=60)
model.train(
    train_df = train_df,
    validation_df = valid_df,
    save_path='./ABW/binary_train/', # where to save the trained model
    epochs=5,
    batch_size=8,
    save_interval=5, #save model every 5 epochs (the best model is always saved in addition)
    num_workers=0, # specify 4 if you have 4 CPU processors, eg., 0 means only the root processor
)

#Let Wandb know that we finished training successfully
wandb.unwatch(model.network)
wandb.finish()



Lorenzo Scala
Senior Marine Scientist
Seiche Ltd
T:
 +44(0)1409 404050
E: ***@***.***
W: https://www.seiche.com
Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 Seiche Ltd is registered in England & Wales. Registered number: 1468514. Registered Office: Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 
 The information contained in this e-mail transmission, and any documents, files or previous e-mail messages attached to it, is privileged and confidential, and solely intended for the use of the individual(s) to whom it is addressed.  If you are not the intended recipient or a person responsible for delivering it to the intended recipient you should not read, copy, distribute or otherwise use the information, and you are hereby notified that any disclosure, copying, distribution or use of any of the information contained in or attached to this message is strictly prohibited.  If you have received this message in error, please notify sender immediately and delete this e-mail and attached documents.

**Reply by canihaveabravo** [reply]
Do you have an example of how to create a label_df when you have a batch of audio files and 1 raven file? The annotation table was created using a BLED in raven processed on a batch of 15-minute wav files. Thank you.

**Reply by sammlapp** [reply]
Hi, thanks for sharing these details. I think you are very close to the solution, and am happy to help with the final details. I've actually never seen a single Raven annotation .txt file containing annotations for multiple audio files (usually we have one annotation file for each audio file). In your example table, for any given annotation (row), how do you know which audio file the annotation corresponds to? 

Specifically, the Begin Time and End Time columns in Raven usually correspond to time since the beginning of the audio file. 

I think that the next step towards using this data to create clip labels is probably to create a column of the table that specifies the corresponding audio file.

**Reply by canihaveabravo** [reply]
In Raven Pro, we open multiple files as a sequence in one window and set the paging to about 180 s. If we then run an automated detector on all files, it returns a single raven selection table. Each annotation is referenced by the corresponding audio file by its Begin time (s) and End time (s) which is incremented continuously from the start of the first audio file in the sequence, rather than the Begin and End of each file as you suggest. We can add a measurement column to display Begin File while shows exactly which audio file contains the annotation. 

Could we then write some code to reference the audio file with begin time and end time of the corresponding annotation perhaps as time (s) since start of first file in sequence?

Thanks very much for your help.

**Reply by canihaveabravo** [reply]
Hello again- We can find the start of each annotation in a file sequence from the File Offset and we can derive the end of an annotation by the sum of File Offset + Delta Time. We can also add a Begin File column to reference which audio file the Offset + Delta corresponds to. With this info at hand would it be possible to write some code to handle audio file sequence annotation tables? Thanks very much for your help and insight. 

Lorenzo

**Reply by canihaveabravo** [reply]
![image](https://github.com/kitzeslab/opensoundscape/assets/12082608/50d7c251-d9be-44cc-8db2-fe830fb176dc)


Image with example Raven measurements showing Begin File, File Offset and Delta Time.

**Reply by sammlapp** [reply]
Hi, yes it will be straightforward to transform this table into the types of tables expected by the BoxedAnnotations.from_raven_files() method. In Python, you could load this table and split it into one annotation file per audio file. Roughly something like this:

Here I'm assuming that "Begin File" is the column containing the audio file name, and that the audio files are in some folder called `/path/to/audio`. 

```python
import pandas as pd
from pathlib import Path

df = pd.read_csv('/path/to/Selection Table.txt',delimiter='\t')

# convert audio file names to full paths
audio_root = Path('/path/to/audio')

# choose where to save annotation files
annotation_root = Path('/path/to/annotations') 
df['Begin File'] = [audio_root / f for f in df['Begin File']]

# convert start time and end time to be relative to audio file start
# adjust the math if this is incorrect
df['Begin Time (s)'] = df['File Offset (s)']
df['End Time (s)'] = df['End Time (s)']+df['File Offset (s)']

# split into one table per audio file. Give it the same name as the audio file
annotation_files = []
for audio_file in df['Begin File']:
    # select rows with this file's annotations
    subset_df = df[df['Begin File']==audio_file]
    annotation_path = annotation_root / (Path(audio_file).stem + '.selections.txt')
    subset_df.to_csv(annotation_path,sep='\t') # saves tab-delimited annotation table
    annotation_files.append(annotation_path) 

# save a table listing each audio and corresponding annotation file path
pd.DataFrame({'audio_file':df['Begin File'],'annotation_file':annotation_files}).to_csv('./audio_and_annotation_files.csv')
```

I haven't run this code because I don't have your files, so it may have some bugs

**Reply by canihaveabravo** [reply]
Thank you very much for providing this code. This has resolved my raven table issue and I'm now able to train my dataset. Fingers crossed for a good outcome. Thanks again.

**Reply by canihaveabravo** [reply]
Hello- Is there a way to instruct the CNN model to ignore unlabelled clips during training? I have a large dataset of audio files that have one hot label clips length 60 s but many of these clips do not contain a label. There might be perhaps 1 annotation in a 15 min file of 15 x 60 s clips. So that's 14 empty clips that pass through the training epoch. Am I understanding that right?

Thanks very much for your advice.

**Reply by louisfh** [reply]
If your 'unlabelled' clips are truly unlabelled - i.e. they've not been annotated and you don't know what they contain, you will want to drop them from the training set. However, if you mean that the unlabelled clips are negatives, i.e. you've listened to them and you know they don't contain any of the sounds of interest you're training your model for, you will want to keep these for training. If your training set is very imbalanced, with many more negatives than positives, you can balance the dataset. A simple way to do this would be to  'upsample' the positives, so you have the same number of positives as negatives.

So if you already have a `labels_dataframe` that looks like this:
| file | start_time        | end_time         | boop | beep | long_boop |
|:-------:|:-------:|:-------------: |:-------------:|:-------:|:-------:|
| file1.wav | 0 | 60 | 1 | 1 | 0 | 
| file1.wav | 60 |   120   | 1 |  0 | 0 |
| file1.wav |120  | 180 | 0 |  0 | 0 |
| file1.wav |180  | 240 | 0 |  0 | 1 |

And the dataframe's index is a multi-index of file, start_time, end_time. (Which will be the case if you've e.g. used our annotations module to make this dataframe)

Here's a simple way of doing this:
```
# find all the rows with at least one non-zero value in one of the columns 
# These contain at least one of your classes of interest and are positives
positives = labels_dataframe[labels_dataframe.sum(axis=1) > 0]

# find all the rows with all 0s in the columns. These are your negatives
negatives = labels_dataframe[labels_dataframe.sum(axis=1) == 0]

# if there are many more negatives than positives, upsample your positives so you have a balanced training set
positives = positives.sample(len(n_negatives), replace=True)

# create a new training set of your negatives and positives
training_set = pd.concat([positives, negatives])
```

You might want to do something different, e.g. if you have tonnes of negatives and your model will take too long to train, you might actually want to downsample your negatives. Or you might want to balance each of the classes if you have multiple classes.

**Reply by canihaveabravo** [reply]
Thanks for your help :)
